Don't create `lastupdate` variables if they are not needed

denisalevi opened the pull request:
For `Synapses` without model equations, the `lastupdate` variable makes up around `~1/3` of the memory demand of a `Synapses` object if I'm not mistaken (`lastupdate`: 8 byte, `preID/postID`: 4 byte each, `delay`: 8 byte). But the `lastupdate` variable is only needed for `event-driven` dynamics or if it is used as identifier in the `Synapses` model equations of the `SynapticPathways` update code.

In this PR I tried to get rid of the `lastupdate`, when its not needed. I'm not sure if I caught all cases (guess the tests wills show). Because of GPU memory limits, this PR would be quite useful for us. But I guess generally also for brian2? Would appreciate some comments.

And in this version I'm checking naively for `lastupdate` being in any of `SynapticPathway` codes. This should be done in a safe way (right now it would also catch `wronglastupdate`). Before I start searching brian for the correct parser function, I was hoping you could just point me to it :)

denisalevi commented:
Modified and added some test. Got a problem when `run_regularly` is using `lastupdate` (see last added test). One would have to add `lastupdate` manually when `run_regularly` uses it. I'm getting the feeling it would have maybe been easier to do the check for `lastupdate` on device side?

mstimberg commented:
I actually stumbled across the unnecessary `lastupdate` variable a few days ago as well, it is not only wasting memory but also the update takes a bit of time.

I wonder whether it makes sense to try to catch all the possible uses, given that they should be corner cases. I guess instead my preferred approach would be:
* The `lastupdate` variable is automatically added when you use event-based synaptic equations
* In all other cases (e.g. [this example](http://brian2.readthedocs.io/en/stable/user/synapses.html#explicit-event-driven-updates)), the user has to explicitly add the variable themselves and add `lastupdate = t` to the end of their `on_pre`/`on_post` statements.

While the latter change is not backwards-compatible, I think it would be ok if we special-case the "unknown identifier" error message for a `lastupdate` variable in synapses, telling the user what to do and/or linking to a page in the documentation.

denisalevi commented:
Your approach seems easier, otherwise there are just a lot of corner cases to catch.

But I just realized that there are also the `N_incoming` and `N_outgoing` variables which are not used most of the time. In standalone mode one could just write them to disk after synapse creation and free their memory. Currently we end up copying them into device memory just for the rare case of someone referencing them in synapses model / update code. And for `N_incoming`/`N_outgoing` one can't just add the variable manually. One could check for it the same way I implemented the `lastupdate` test here, but then we would again check for corner cases. Any idea how to go about that?

mstimberg commented:
When we introduced `N_incoming` and `N_outgoing`, we discussed the memory waste given that they are not used most of the time. We decided to not worry, given that they are only stored once per neuron, not once per synapse (they are accessible as synaptic variables, but that's just indexing magic). 

In principle it might be useful to have a general mechanism to just copy over variables to device code that are actually used on the device side? You wouldn't have to parse any code for that, you could look at the `variables` dictionaries of all code objects.

denisalevi commented:
> When we introduced N_incoming and N_outgoing, we discussed the memory waste given that they are not used most of the time. We decided to not worry, given that they are only stored once per neuron, not once per synapse (they are accessible as synaptic variables, but that's just indexing magic).

Yes. Forgot that when I wrote my last comment. I agree.

> In principle it might be useful to have a general mechanism to just copy over variables to device code that are actually used on the device side? You wouldn't have to parse any code for that, you could look at the variables dictionaries of all code objects.

Yes I would love to refactor that variable initialisation code and have a good mechanism for host to device copies. It just grew out of Konrads code and I just added more and more exceptions or started deleting variables that are initialised and not used etc... its quite messy. But I think we don't have time for such a change right now. And for `lastupdate` such a variable inspection mechanism wouldn't work as it is used in the update code.

So the implementation here (which I am currently using) breaks brian2genn performance (see brian-team/brian2genn#64). If there is some decision to remove `lastupdate`, that will need some change in `brian2genn` I guess.

How about implementing your suggestion of getting rid of `lastupdate` entirely for not event-driven synapses? The check is very easily implemented, I guess it would be more about documenting it (and well, the decision to do it). So what do you think? I would love to see `lastupdate` gone for our result benchmarks :)

mstimberg commented:
This has been made obsolete by PR #1003. Thanks for brining this up @denisalevi !
